include/beast/core/impl/buffers_adapter.ipp

```diff
@@ -231,10 +220,17 @@ public:
     {
         using boost::asio::buffer_cast;
         using boost::asio::buffer_size;
+        auto last = ba_->begin_;
+        for(;;)
+        {
+            auto next = std::next(last);
+            if(next == ba_->end_)
+                break;
+            last = next;
+        }
         return value_type{buffer_cast<void*>(*it_),
-            it_ == std::prev(ba_->end_) ?
-                ba_->out_end_ : buffer_size(*it_)} +
-                    (it_ == ba_->out_ ? ba_->out_pos_ : 0);
+            it_ == last ? ba_->out_end_ : buffer_size(*it_)} +
+                (it_ == ba_->out_ ? ba_->out_pos_ : 0);
```

Comment:
This looks like an excellent place to write a generic algorithm to find the last dereferenceable in a range.  It can be easily optimized to O(1) if the iterator supports bidirectional (which it doesn't have to but it might):
```
namespace detail
{

template <class FwdIt>
FwdIt
find_last(FwdIt first, FwdIt last, std::forward_iterator_tag)
{
    for(;;)
    {
        auto next = std::next(first);
        if (next == last)
            break;
        first = next;
    }
    return first;
}

template <class BidIt>
BidIt
find_last(BidIt, BidIt last, std::bidirectional_iterator_tag)
{
    return std::prev(last);
}

}  // detail

template <class FwdIt>
FwdIt
find_last(FwdIt first, FwdIt last)
{
    assert(first != last);
    return detail::find_last(first, last,
                             typename std::iterator_traits<FwdIt>::iterator_category{});
}
```